src/state/routing/useRoutingAPITrade.ts

```diff
@@ -0,0 +1,175 @@
+ * query should be skipped.
+ */
+function useRoutingAPIArguments({
+  tokenIn,
+  tokenOut,
+  amount,
+  type,
+}: {
+  tokenIn: Currency | undefined
+  tokenOut: Currency | undefined
+  amount: CurrencyAmount<Currency> | undefined
+  type: 'exactIn' | 'exactOut'
+}) {
+  if (!tokenIn || !tokenOut || !amount || tokenIn.equals(tokenOut)) {
+    return undefined
+  }
+
+  return {
+    tokenInAddress: tokenIn.wrapped.address,
+    tokenInChainId: tokenIn.chainId,
+    tokenOutAddress: tokenOut.wrapped.address,
+    tokenOutChainId: tokenOut.chainId,
+    amount: amount.quotient.toString(),
+    type,
+  }
+}
+
+export function useRoutingAPITradeExactIn(amountIn?: CurrencyAmount<Currency>, currencyOut?: Currency) {
+  const queryArgs = useRoutingAPIArguments({
+    tokenIn: amountIn?.currency,
+    tokenOut: currencyOut,
+    amount: amountIn,
+    type: 'exactIn',
+  })
+
+  const { isLoading, isError, data } = useGetQuoteQuery(queryArgs ?? skipToken, {
+    pollingInterval: ms`10s`,
+    refetchOnFocus: true,
+  })
+
+  const quoteResult = useFreshData(data, Number(data?.blockNumber) ?? 0)
+
+  const routes = useMemo(
+    () => computeRoutes(amountIn?.currency, currencyOut, quoteResult),
+    [amountIn, currencyOut, quoteResult]
+  )
+
+  return useMemo(() => {
+    if (!amountIn || !currencyOut) {
+      return {
+        state: V3TradeState.INVALID,
+        trade: null,
+      }
+    }
+
+    if (isLoading && !quoteResult) {
+      // only on first hook render
+      return {
+        state: V3TradeState.LOADING,
+        trade: null,
+      }
+    }
+
+    const amountOut =
+      currencyOut && quoteResult ? CurrencyAmount.fromRawAmount(currencyOut, quoteResult.quote) : undefined
+
+    if (isError || !amountOut || !routes || routes.length === 0 || !queryArgs) {
+      return {
+        state: V3TradeState.NO_ROUTE_FOUND,
+        trade: null,
+      }
+    }
+
+    const trade = Trade.createUncheckedTradeWithMultipleRoutes<Currency, Currency, TradeType.EXACT_INPUT>({
+      routes,
+      tradeType: TradeType.EXACT_INPUT,
+    })
+
+    const gasPriceWei = BigNumber.from(quoteResult?.gasPriceWei)
+    const gasUseEstimate = BigNumber.from(quoteResult?.gasUseEstimate)
+
+    return {
+      // always return VALID regardless of isFetching status
+      state: V3TradeState.VALID,
+      trade,
+      gasPriceWei,
+      gasUseEstimate,
+    }
+  }, [amountIn, currencyOut, isLoading, quoteResult, isError, routes, queryArgs])
+}
+
+export function useRoutingAPITradeExactOut(currencyIn?: Currency, amountOut?: CurrencyAmount<Currency>) {
+  const queryArgs = useRoutingAPIArguments({
+    tokenIn: currencyIn,
+    tokenOut: amountOut?.currency,
+    amount: amountOut,
+    type: 'exactOut',
+  })
+
+  const { isLoading, isError, data } = useGetQuoteQuery(queryArgs ?? skipToken, {
+    pollingInterval: ms`10s`,
+    refetchOnFocus: true,
+  })
+
+  const quoteResult = useFreshData(data, Number(data?.blockNumber) ?? 0)
```

Comment:
`Number(undefined) ?? 0` == `NaN`, not what we want. it's inconsequential in this case but could cause a bug

Reply:
good catch! `|| 0` is a good alternative yeah? Given `NaN` cannot be evaluated to `true`